### PR TITLE
docs: RFC Mutually Exclusive Pods Uniformly Distributed Scheduling

### LINKED ISSUE #1418: node selection: One supper large node with many small size nodes
### Description

**Observed Behavior**:
I have 10 pending pods:
- **pod1**: 1c1g requests, with anti-affinity; cannot be scheduled on the same node as pod10 and pod9.
- **pod2 ~ pod8**: 1c1g requests; no anti-affinity is configured.
- **pod9**: 1c1g requests, with anti-affinity; cannot be scheduled on the same node as pod1 and pod10.
- **pod10**: 1c1g requests, with anti-affinity; cannot be scheduled on the same node as pod1 and pod9.

In Karpenter, it creates three nodes, with the first node being super large and the remaining two being very small:
- **node1**: c7a.4xlarge, 16c32g (8Pods)
- **node2**: c7a.large, 2c4g (1Pod)
- **node3**: c7a.large, 2c4g (1Pod)

**Expected Behavior**:
I want the resources of the three nodes to be evenly distributed, like:
- **node1**: c7a.4xlarge, 8c16g (4Pod)
- **node2**: c7a.xlarge, 4c8g (3Pod)
- **node3**: c7a.xlarge, 4c8g (3Pod)

In this situation, the cluster will be more stable (e.g., draining one node will not cause most pods to be rescheduled). I observed that Karpenter attempts to distribute the pods across all nodes:
[Scheduler Code](https://github.com/kubernetes-sigs/karpenter/blob/c4be45e04079f0d9313150b3ae7b5313132b0e36/pkg/controllers/provisioning/scheduling/scheduler.go#L259)

However, if the pods are not sorted and those with anti-affinity are not scheduled first in the simulation process, the problem will occur as described.

**Versions**:
- Chart Version: v0.36.0
- Kubernetes Version (`kubectl version`): v1.29.0

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by helen-frank**
Fixes #1418 

**Description**

**How was this change tested?**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by helen-frank** [comment]
Request for review of RFC
/cc @ellistarn @jwcesign

**Comment by helen-frank** [comment]
ping @ellistarn

**Comment by helen-frank** [comment]
/cc @njtran @ellistarn

Please take a look

File: designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+- **node1**: c7a.4xlarge, 8c16g (4Pod)
+- **node2**: c7a.xlarge, 4c8g (3Pod)
+- **node3**: c7a.xlarge, 4c8g (3Pod)
+
+In this situation, the cluster will be more stable (e.g., draining one node will not cause most pods to be rescheduled).
+
+## Recommended Solution
+
+By sorting in byCPUAndMemoryDescending, pods with the same specifications (cpu or memory request equal) and mutually exclusive pods (with PodAntiAffinity or TopologySpreadConstraints) are prioritized to be scheduled first.
+
```
Comment (jwcesign) [inline designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md:32]: Can you explain the theory details here?

File: designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+**Expected Behavior**:
+I want the resources of the three nodes to be evenly distributed, like:
+
+- **node1**: c7a.4xlarge, 8c16g (4Pod)
+- **node2**: c7a.xlarge, 4c8g (3Pod)
+- **node3**: c7a.xlarge, 4c8g (3Pod)
+
+In this situation, the cluster will be more stable (e.g., draining one node will not cause most pods to be rescheduled).
+
+## Recommended Solution
```
Comment (jwcesign) [inline designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md:29]: Proposal

File: designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+By sorting in byCPUAndMemoryDescending, pods with the same specifications (cpu or memory request equal) and mutually exclusive pods (with PodAntiAffinity or TopologySpreadConstraints) are prioritized to be scheduled first.
+
+Code implementation: [pull 1548](https://github.com/kubernetes-sigs/karpenter/pull/1548)
+
+### Test Results
+
```
Comment (jwcesign) [inline designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md:36]: What will be the impact if a user upgrades to a version that includes this feature? 

Will the existing nodes need to be recreated according to the new logic?

Comment (helen-frank) [reply designs/mutually-exclusive-pods-uniformly-distributed-scheduling.md:38]: The new scheduling results are more balanced, and that's about it. No changes are needed for that.

**Comment by helen-frank** [comment]
@njtran @ellistarn @jonathan-innis Recently, I have some free time. Can we continue this work

**Pull request closed without merging**